### PR TITLE
Fixes #10677: Datasources plugin call continuously datasource url even after delete

### PR DESCRIPTION
**Pull request by fanf**
https://www.rudder-project.org/redmine/issues/10677

That pull request correct a nasty race condition. I *was not able* to reproduce it live, but the race is clearly there. 
I could just have added synchronized on updateDataSourceScheduler, but it would have been bugs wainting to flourish, so I prefered to encapsulate all the must-be synchronized methods dealing with our mutable var in a synchronized object. 

To sum up: mutable var are bad. Mutable vars combined with threads are a can of worms, the worms being able to fly and kill any life on a touch. And yes, you have to unknot the thing.

### CODE REVIEW

**Comment by fanf** [comment]
PR rebased

File: src/main/scala/com/normation/plugins/datasources/Repository.scala [hunk 1]
```diff
@@ -199,8 +215,6 @@ class DataSourceRepoImpl(
 
   }
   private[this] def updateDataSourceScheduler(source: DataSource, delay: Option[FiniteDuration]): Unit = {
```
Comment (fanf) [inline src/main/scala/com/normation/plugins/datasources/Repository.scala:217]: So, the problem was most likely here: that method is not synchronized, and was used on non sync method. So what could have happened: 
[context $1]
- stop source.id $0, 
[switch context to $2]
- stop source.id $0, 
- create dss $2
- save dss $2
- start dss $2
[switch context $1]
- create dss $1
- save dss $1
- start dss $1

So now, both $1 and $2 run, but only $1 is know for rudder. This is bad.

File: src/main/scala/com/normation/plugins/datasources/Repository.scala [hunk 2]
```diff
@@ -242,7 +256,7 @@ class DataSourceRepoImpl(
    * on update, we need to stop the corresponding optionnaly existing
    * scheduler, and update with the new one.
    */
-  override def save(source : DataSource) : Box[DataSource] = synchronized {
+  override def save(source : DataSource) : Box[DataSource] = {
```
Comment (ncharles) [inline src/main/scala/com/normation/plugins/datasources/Repository.scala]: i *think* it shold stay synchronized (but not 100% sure)

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

File: src/main/scala/com/normation/plugins/datasources/Repository.scala [hunk 3]
```diff
@@ -161,7 +161,31 @@ class DataSourceRepoImpl(
...
+   * Be careful, ALL modification to datasource must be synchronized
+   */
+  private[this] object datasources extends AnyRef {
+    private[this] var internal = Map[DataSourceId, DataSourceScheduler]()
+    // utility methods on datasources
+    // stop a datasource - must be called when the datasource still in "datasources"
+    private[this] def stop(id: DataSourceId) = {
+      DataSourceLogger.debug(s"Stopping data source with id '${id.value}'")
+      internal.get(id) match {
+        case None      => s"Data source with id ${id.value} was not found running"
```
Comment (ncharles) [inline src/main/scala/com/normation/plugins/datasources/Repository.scala]: this text doesn't go anywhere, unless i'm mistaken. Shouldn't it go to a logger ?

Comment (fanf) [reply src/main/scala/com/normation/plugins/datasources/Repository.scala]: yes :)

**Comment by fanf** [comment]
PR rebased
